Update taint feature design doc and user manual (#1632)
FoundationDB/fdb-kubernetes-operator: docs/design/node_taint_eviction.md

## Metadata

* Authors: @johscheuer, @xumengpanda
* Created: 2023-02-14
* Updated: 2023-03-08

> * Updated: 2023-05-11

Reviewer: Remove the line above. We only want to keep 1 updated statement.

## Background

The operator is currently able to detect failures of Pods and replace them if required, the replacement is happening based on the [conditions that need replacement](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/v1.14.0/api/v1beta2/foundationdbcluster_types.go#L65).

> One of the missing pieces in failure mitigation is to evict Pods from tainted Nodes.

Author: It's the same reason of using uppercase for Pods.

In Kubernetes it's common to use [taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration) to indicate potential problems with a node or planned maintenance.
This information could be consumed by the operator to proactively replace Pods that are running on Nodes with a given set of taints.

## General Design Goals

The goal of this design is to implement a mechanism that allows the operator to replace Pods that are running on Nodes with a given set of taints.
It's expected that the operator will detect the Pods running on tainted Nodes in a timely manner, either by receiving an event triggering the reconciliation loop or by periodic reconciliation events that happen every 10h by default. Once the operator marks Pods on tainted Nodes as to-be-replaced, the operator will re-use the replacement logic to replace those Pods. The actual replacement speed depends on the speed of excluding a Pods, the number of Pods to replace, and the allowed maximum number of inflight replacements (defined by `automationOptions.replacements.maxConcurrentReplacements`).

## Current Implementation

> The operator does not have this feature. It does not have access to list Node information. It does not consider taint-related events in Pod's condition.

Author: I think this is clearer and simpler.

> The operator does replace failed ProcessGroups in the [Replace Failed ProcessGroups reconciler](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/controllers/replace_failed_process_groups.go). It can extend it to support this feature.

Author: I think the changed version is clearer and simpler.

## Proposed Design

### Design Requirements

> **Requirement 1.** The feature can be enabled and disabled by users, because (1) when someone wants to run the operator inside a k8s cluster without Node access, they should be able to bypass this feature; (2) when we notice problems during roll-out of this feature, we need to disable it online.

Reviewer: What is the benefit of changing the layout here? What do you mean by "we need to disable it online"? Why is this a requirement?

Author: we need to be able to disable it online (ie., on-the-fly)

Reviewer: Not sure why this has to be "online" and a redeployment of the operator with the according settings to disable wouldn't be good enough.

Author: I see. redeployment also works. let me remove `online`.

> The feature can be disabled by option `automationOptions.replacements.taintReplacementOptions = {}`.

Reviewer: This is not a requirement but an implementation detail, this should be in the user manual.

Author: This is the API definition/design for users.
During our discussion in implementation PR, we debated on "should we support using {key=*, duration=-1}" to disable the entire taint feature.  If the design defines it clearly,  the implementation will have less ambiguity.

Reviewer: If this is intended for users we should make this example in `yaml` in the context of the `FoundationDBClusterSpec` and not a one line go code example as most (if not all) users will be creating and managing clusters using yaml files with the `FoundationDBCluster` definition.

Author: add a yaml example

> **Requirement 2.** The feature allows users to configure a set of specific taint keys to detect and replace. This is useful in a multi-tenant k8s environment where FDB operator only wants to react to a subset of taints used in the k8s cluster.

Author: :) Sure.

> **Requirement 2.** The feature allows users to configure a set of specific taint keys to detect and replace. This is useful in a multi-tenant k8s environment where FDB operator only wants to react to a subset of taints used in the k8s cluster.

Author: I was thinking about k8s or C* team may use taint for their own purpose and we don't want to replace Pods for that particular taint.

Users can configure the taint feature in the new field `taints` inside `automaticReplacementOptions`. The `taints` field is a list of `(key, durationInSeconds)` pairs, which defines operator's behavior on the taint key.

> Users can use `automaticReplacementOptions.taintReplacementTimeSeconds` to control how fast operator should replace Pods on tainted Nodes. Users can ask operator to replace Pods on tainted Nodes much faster than the normal replacement by setting a smaller value of `taintReplacementTimeSeconds`.

Reviewer: `Users can ask operator to replace Pods on tainted Nodes`, how does a user "asks" the operator to replace Pods on a tainted node? Can we stick to a more technical language? I would prefer the word "configure" and not "ask".

  enableNodeFailureDetection: true
  taints:
  - key: "example.org/maintenance"

>     durationInSeconds: 7200 # Mark Pods to be replaced after this taint key has been on the Node for at least 7200 seconds (i.e., 2 hours).

Reviewer: What is the benefit here with those changes? Adding `taintReplacementTimeSeconds` makes sense, but I'm not sure about the value of adding yet another example?

Author: it's to show we can add a list of different taint keys and wildcard taint key.

Reviewer: ok.

  - key: "example.org/disconnected"
    durationInSeconds: 1200 # Mark Pods to be replaced after this taint key has been on the Node for at least 1200 seconds (i.e., 20 mins).
  - key: "*" # The wildcard will match all taint keys
    durationInSeconds: 3600 # Mark Pods to be replaced after a taint key, which does not have exact match in the taints list, has been on the Node for at least 3600 seconds.
  taintReplacementTimeSeconds: 1800 # A Pod marked to be replaced due to taints must stay in the to-be-replaced state for 1800 seconds before it is automatically replaced.
```

In the above configuration, it takes at least 3000 seconds (1200 + 1800) for the operator to replace a Pod on a Node with taint key `example.com/disconnected`, because it takes `durationInSeconds`, i.e., 1200 seconds, to mark the Pod to be replaced and it takes another `taintReplacementTimeSeconds`, i.e., 1800 seconds, to start replacing the Pod.

**Requirement 3.** The feature lists the taint status for process groups. This allows users to have a holistic view of which Pods in a FDB cluster are running on tainted Nodes and being replaced.

> ### Implementation
> We need to make the following four type of changes:

Reviewer: ```suggestion
### Implementation

We need to make the following four type of changes:
```

> **CRD change.** We need to extend the `automaticReplacementOptions` with the new fields `taints` and `taintReplacementTimeSeconds`

Reviewer: This information is already implicitly given as we stated above the we are extending the `AutomaticReplacementOptions`.

Author: this is to echo the design requirement.

> **Detect Node taint and change Pod condition.**  We can piggy-back the existing `validateProcessGroup()` function in `update_status` reconciler to detect if a Pod is running on a tainted Node and mark the Pod's taint-related condition accordingly.

Reviewer: ```suggestion
**Detect Node taint and change ProcessGroup condition.**  We can extend the existing `validateProcessGroup()` function in `update_status` reconciler to detect if a Pod is running on a tainted Node and mark the ProcessGroup taint-related condition accordingly.
```

We are not changing any Pod condition. We change the `ProcessGroup` condition, those are different things.

I'm not sure what "type" of change this is but I would have expected something like "code changes", "CRD changes", "documentation changes" etc. as "type" of change.

Author: This item is a procedure, which is like algorithm. 
The CRD changes are like data-structure change.

A Pod's Node can be listed by
```
node = &corev1.Node{}

> err := r.Get(ctx, client.ObjectKey{Name: pod.Spec.NodeName}, node)

Reviewer: What is the value with this statement? This is a basic example for how to use the controller runtime, can we remove this? What is the value of adding this example here?

```

> **Status of Pod taint-related condition.** We will add two  [ProcessGroupConditionType](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/api/v1beta2/foundationdbcluster_types.go#L673): `NodeTaintDetected`, which represents a Pod's Node is tainted but not long enough for operator to replace it, and `NodeTaintReplacing`, which represents a Pod whose Node has been tainted for longer-than-configured time and the operator should replace it.

Reviewer: ```suggestion
**Status of ProcessGroup taint-related condition.** We will add two [ProcessGroupConditionType](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/api/v1beta2/foundationdbcluster_types.go#L673):

- `NodeTaintDetected` represents a ProcessGroup Node is tainted, but not long enough to replace it.
- `NodeTaintReplacing` represents a ProcessGroup whose Node has been tainted for longer-than-configured time and the ProcessGroup should be replaced.
```

Same as above in this context we are talking about `ProcessGroup`s and not `Pod`s. I personally would prefer a listing instead of a long sentence.

> **Replace Pods with `NodeTaintReplacing` conditioin.** We can piggy-back existing replacement logic to replace Pods with `NodeTaintReplacing` condition. We can do this by adding `NodeTaintReplacing` to [`conditionsThatNeedReplacement` list](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/d38d6bc7abf7764215976b75d581c933280389f8/api/v1beta2/foundationdbcluster_types.go#L65-L67)

Reviewer: ```suggestion
**Replace ProcessGroup with `NodeTaintReplacing` condition.** We can extend the existing replacement logic to replace ProcessGroups with `NodeTaintReplacing` condition. We can do this by adding the `NodeTaintReplacing` condition to [`conditionsThatNeedReplacement` list](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/d38d6bc7abf7764215976b75d581c933280389f8/api/v1beta2/foundationdbcluster_types.go#L65-L67)
```

Same as above in this context we are talking about ProcessGroups and not Pods.

> ## Limitations
> The operator will only detect tainted Nodes when its reconciliation is triggered periodically. This may increase the delay of detecting and replacing Pods on tainted Nodes.

Reviewer: ```suggestion
## Limitations

The operator will only detect tainted Nodes when its reconciliation is triggered. This may increase the delay of detecting and replacing Pods on tainted Nodes.
```

The periodically statement is wrong here. The operator will also (potentially) detected this condition if a reconciliation was triggered by an unrelated event, saying that the tainted nodes are only detected by the periodic reconciliation is wrong. We can add a statement here that node changes are not triggering any reconciliation.

> The limitation can be resolved by letting the operator watch node events. The controller-runtime already supports to let the operator watch resources that are not managed by itself, see [Watching Externally Managed Resources](https://book.kubebuilder.io/reference/watching-resources/externally-managed.html).

Reviewer: ```suggestion
The limitation can be resolved by letting the operator watch node events. The controller-runtime already supports to let the operator watch resources that are not managed by itself, see [Watching Externally Managed Resources](https://book.kubebuilder.io/reference/watching-resources/externally-managed.html).
```

You could also reference: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/1629, so users can see if the feature is implemented.

> Because a k8s cluster may have a high volume of node events, we must make sure we get the right signals and not overload the operator by excessive events.

Reviewer: This is a limitation? I would rather see this as a requirement. Is this statement part of the sentence above?

Author: yes

Author: removed this sentecne
